File: vehicle/haskell/src/Submarine.hs
module Submarine (Submarine (..)) where

import Vehicle (Vehicle)

class Vehicle submarineType => Submarine submarineType where
  getEngine :: submarineType -> String
  getEngine _ = "gasoline engine"

  dive :: submarineType -> IO ()
  dive submarine = do
    putStrLn ("Submarine engine" <> getEngine submarine)
    putStrLn "&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&"

  simulate :: submarineType -> IO ()
  simulate = dive

File: vehicle/rust/src/submarine.rs
use crate::Vehicle;

pub trait Submarine: Vehicle {
  fn get_engine(&self) -> String {
    String::from("gasoline engine")
  }

  fn dive(&self) {
    println!("Submarine engine: {}", Submarine::get_engine(self));
    println!("&&&&&&&&&&&&&&&&&&&&&&&&&&");
  }

  fn simulate(&self) {
    self.dive();
  }
}
